### DISCUSSION TITLE
Infinitely-sized quad-tree implementation question

### DISCUSSION THREAD

**Discussion by aecsocket** [jrouwe/JoltPhysics#515] [Q&A] [answered]
I've been trying to understand how the broad phase algorithm works in Jolt, but I don't understand how the quad-tree implementation works. In other physics engines like PhysX, a bounding volume hierarchy is used instead, which I'm able to understand better. However using a quad-tree seems to pose some problems which I don't understand how they are solved in the code:

- Are individual points (a body's position) or an AABB (a body's bounds) inserted into the tree? What happens if that body spans multiple quadrants of the tree?
- How does the quad-tree scale to an infinitely sized world? From the implementations I can see online, you need to know the entire size of the tree's root node first, but this can't be the case for Jolt.
  - What happens then when a lot of bodies are inserted around (10, 10, 10), let's say between (5, 5, 5) to (15, 15, 15), then another body is inserted much further along at (1000, 10, 10)?
- How is adding multiple bodies add once (through AddBodiesPrepare/AddBodiesFinalize) handled? From the code it looks like it builds a separate tree and calculates the total bounds of that, but how is this then added to the whole tree?

**Comment by jrouwe** [comment]
Hello,

Please take a look at my GDC22 talk: https://gdcvault.com/play/1027560/Architecting-Jolt-Physics-for-Horizon (slides only) or https://gdcvault.com/play/1027891/Architecting-Jolt-Physics-for-Horizon if you have a GDC vault subscription

**Reply by aecsocket** [reply]
Thanks for the links. Unfortunately I don't have a GDC subscription so I can't watch the talk, but I'm sure it's very useful! The slides do answer most of my questions, although I have two more:

- What made you use a quad tree instead of a BVH? Is there some sort of performance or implementation difference that I don't see?
- What terminating condition do you use when splitting your quad tree?

Thanks in advance!

**Comment by jrouwe** [comment, marked as answer]
1. The quad tree in this case is a BVH, it keeps the bounding box for each of the 4 children
2. Each node can have up to 4 children/bodies so the termination condition is that there is 1 body left

**Reply by aecsocket** [reply]
That makes a lot more sense now, thanks for the explanation!
